temporalio/worker/_workflow_instance.py

```diff
@@ -414,6 +419,100 @@ def _apply_cancel_workflow(
             # this cancellation to the next iteration of the event loop.
             self.call_soon(self._primary_task.cancel)
 
+    def _apply_do_update(
+        self, job: temporalio.bridge.proto.workflow_activation.DoUpdate
+    ):
+        # Run the validator & handler in a task. Everything, including looking up the update definition, needs to be
+        # inside the task, since the update may not be defined until after we have started the workflow - for example
+        # if an update is in the first WFT & is also registered dynamically at the top of workflow code.
+        async def run_update() -> None:
+            command = self._add_command()
+            command.update_response.protocol_instance_id = job.protocol_instance_id
+            try:
+                defn = self._updates.get(job.name) or self._updates.get(None)
+                if not defn:
+                    known_updates = sorted([k for k in self._updates.keys() if k])
+                    raise RuntimeError(
+                        f"Update handler for '{job.name}' expected but not found, and there is no dynamic handler. "
+                        f"known updates: [{' '.join(known_updates)}]"
+                    )
+                args = self._process_handler_args(
```

Comment:
Based on our discussion the other day, we have to re-process args again after validation to prevent mutation from having an effect during validator

Reply:
Fixed